### DISCUSSION TITLE
How to handle the grainid beyond the number of grains

### DISCUSSION THREAD

**Discussion by kun-Jiang** [mtex-toolbox/mtex#2131] [Ask Anything] [answered]
Hello everyone,

I'm currently working on calculating the dislocation density around the crack path, and I've been following a tutorial to achieve this. However, I encountered an error when running the following command:
Command:
`
ipfKey.oriRef = grains(ebsd('Ni (Nickel)').grainId).meanOrientation;
`
The error message states:
`
Out of range; index must be a positive integer or logical.
`
I checked the code and find that the maximum id obtained from `ebsd('Ni (Nickel)').grainId` is 1503, but there is only 477 grains. This inconsistency has left me puzzled. So how did this problem arise?

Following is my code and structure:
![image](https://github.com/mtex-toolbox/mtex/assets/53817143/558809f1-7bd1-4295-b980-4ffdd4e11c83)
![image](https://github.com/mtex-toolbox/mtex/assets/53817143/f58afc9f-6139-4f51-9888-90ec235c9a08)

**Comment by zmichels** [comment]
Does it work if you add 'id' in the command?

Like:

ipfKey.oriRef = grains('id', ebsd('Ni (Nickel)').grainId).meanOrientation

**Reply by kun-Jiang** [reply]
Thanks, zmichels, but it still doesn't work. This is the error message
`
No data with the specified ids in the data set
`

**Reply by zmichels** [reply, marked as answer]
What about if you add (after filtering out some grains and before the problematic line):

ebsd = ebsd(grains);

**Reply by kun-Jiang** [reply]
Thanks, zmichels. That works well. So what is the meaning of this command?

**Reply by zmichels** [reply]
@kun-Jiang I admit I'm not entirely sure why it's necessary, but somehow the id affiliations are confused when there are EBSD points for grains that have been removed from the grainset. Whenever I filter out grains I do this to avoid similar issues you had. I didn't notice the first time looking through the code tho.
